File: Musicefy.Core/Models/MusicFile.cs
using System;

namespace Musicefy.Core.Models
{
    public class MusicFile
    {
        // Unique identifier
        public string Id { get; set; }

        // File path on disk (local) or remote URI
        public string FilePath { get; set; }

        // Core metadata
        public string Title { get; set; }
        public string Artist { get; set; }
        public string Album { get; set; }
        public int Year { get; set; }
        public string Genre { get; set; }
        public TimeSpan Duration { get; set; }
        public int TrackNumber { get; set; }

        // Extended metadata
        public int Bitrate { get; set; }            // kbps
        public long FileSize { get; set; }          // bytes
        public string Lyrics { get; set; }          // raw lyrics text
        public string CoverPath { get; set; }       // album art image path

        // Source info
        public string SourceUri { get; set; }       // streaming/local source
        public string SourceType { get; set; }      // e.g. Local, Subsonic, Spotify

        // User interaction
        public int PlayCount { get; set; }
        public DateTime LastPlayed { get; set; }
        public bool IsFavourite { get; set; }
        public bool IsDownloaded { get; set; }

        // Constructors
        public MusicFile()
        {
            Id = Guid.NewGuid().ToString();
        }

        public MusicFile(
            string title,
            string artist,
            string album = "",
            int year = 0,
            string sourceUri = null,
            string filePath = null,
            string genre = null,
            TimeSpan duration = default,
            int trackNumber = 0,
            string sourceType = "Local",
            int bitrate = 0,
            long fileSize = 0,
            string lyrics = null,
            string coverPath = null)
        {
            Id = Guid.NewGuid().ToString();
            Title = title;
            Artist = artist;
            Album = album;
            Year = year;
            SourceUri = sourceUri;
            FilePath = filePath ?? sourceUri ?? string.Empty;
            Genre = genre;
            Duration = duration;
            TrackNumber = trackNumber;
            SourceType = sourceType;
            Bitrate = bitrate;
            FileSize = fileSize;
            Lyrics = lyrics;
            CoverPath = coverPath;
            PlayCount = 0;
            LastPlayed = DateTime.MinValue;
            IsFavourite = false;
            IsDownloaded = false;
        }

        // Methods
        public void MarkPlayed()
        {
            PlayCount++;
            LastPlayed = DateTime.Now;
        }

        public void ToggleFavourite() => IsFavourite = !IsFavourite;

        public override bool Equals(object obj)
        {
            return obj is MusicFile other &&
                   string.Equals(FilePath, other.FilePath, StringComparison.OrdinalIgnoreCase);
        }

        public override int GetHashCode()
        {
            return FilePath != null ? StringComparer.OrdinalIgnoreCase.GetHashCode(FilePath) : 0;
        }

        public override string ToString()
        {
            return string.IsNullOrEmpty(Artist) ? Title : $"{Title} - {Artist}";
        }
    }
}

File: src/core/models/MusicFile.cpp
// MusicFile.cpp

#include "MusicFile.h"

#include <QUuid>

namespace mf::core::models {

MusicFile::MusicFile()
    : id_(QUuid::createUuid().toString(QUuid::WithoutBraces))
    , lastPlayed_(QDateTime::fromSecsSinceEpoch(0))
{
}

MusicFile::MusicFile(QString title,
                     QString artist,
                     QString album,
                     int year,
                     QString sourceUri,
                     QString filePath,
                     QString genre,
                     std::chrono::seconds duration,
                     int trackNumber,
                     QString sourceType,
                     int bitrate,
                     qint64 fileSize,
                     QString lyrics,
                     QString coverPath)
    : id_(QUuid::createUuid().toString(QUuid::WithoutBraces))
    , filePath_(filePath.isEmpty() ? sourceUri : filePath)
    , title_(std::move(title))
    , artist_(std::move(artist))
    , album_(std::move(album))
    , year_(year)
    , genre_(std::move(genre))
    , duration_(duration)
    , trackNumber_(trackNumber)
    , bitrate_(bitrate)
    , fileSize_(fileSize)
    , lyrics_(std::move(lyrics))
    , coverPath_(std::move(coverPath))
    , sourceUri_(std::move(sourceUri))
    , sourceType_(std::move(sourceType))
    , playCount_(0)
    , lastPlayed_(QDateTime::fromSecsSinceEpoch(0))
    , isFavourite_(false)
    , isDownloaded_(false)
{
}

void MusicFile::markPlayed() {
    ++playCount_;
    lastPlayed_ = QDateTime::currentDateTime();
}

QString MusicFile::coverUrl() const {
    if (!coverUrl_.isEmpty()) return coverUrl_;
    return coverPath_;
}

bool MusicFile::operator==(const MusicFile& other) const {
    return QString::compare(filePath_, other.filePath_, Qt::CaseInsensitive) == 0;
}

QString MusicFile::toDisplayString() const {
    if (artist_.isEmpty()) {
        return title_;
    }
    return title_ + QStringLiteral(" - ") + artist_;
}

} // namespace mf::core::models
